docs/ai/reference.rst

```diff
@@ -0,0 +1,520 @@
+            distance := 0.8560434728860855
+          ),
+        }
+
+
+HTTP endpoints
+==============
+
+
+RAG
+---
+
+``POST``: ``https://<edgedb-host>:<port>/branch/<branch-name>/ai/rag``
+
+Responds with text generated by the specified text generation model.
+
+
+Request body
+^^^^^^^^^^^^
+
+The request body should be JSON.
+
+* ``model`` (string, required): The name of the text generation model to use.
+
+* ``query`` (string, required): The query string use as the basis for text
+  generation.
+
+* ``context`` (object, required): Settings that define the context of the
+  query.
+
+  * ``query`` (string, required): Specifies the query to determine the relevant
+    objects and index to serve as context for text generation.
+
+  * ``variables`` (object, optional): A dictionary of variables for use in the
+    context query.
+
+  * ``globals`` (object, optional): A dictionary of globals for use in the
+    context query.
+
+  * ``max_object_count`` (int, optional): Maximum number of objects to return;
+    default is 5.
+
+* ``stream`` (boolean, optional): Specifies whether the response should be
+  streamed. Defaults to false.
+
+* ``prompt`` (object, optional): Settings that define a prompt, overriding the
+  default prompt.
+
+  You may specify an existing prompt by its ``name`` or ``id``, or you may
+  define a custom prompt inline by sending an array of objects, each containing
+  a ``role`` and ``content``. Choose to specify any one of these properties, or
+  omit the ``prompt`` property to use the default prompt:
+
+  * ``name`` (string, optional): The ``name`` of an existing custom prompt to
+    use.
+
+  * ``id`` (string, optional): The ``id`` of an existing custom prompt to use.
+
+  * ``custom`` (array of objects, optional): Custom prompt messages, each
+    containing a ``role`` and ``content``.
+
+
+Response
+^^^^^^^^
+
+**Example successful response**
+
+- **HTTP status**: 200 OK
+- **Content-Type**: application/json
+- **Body**:
+
+  .. code-block:: json
+
+      {"response": "The sky on Mars is red."}
+
+**Example error response**
+
+- **HTTP status**: 400 Bad Request
+- **Content-Type**: application/json
+- **Body**:
+
+  .. code-block:: json
+
+      {
+        "message": "missing required 'query' in request 'context' object",
+        "type": "BadRequestError"
+      }
+
+
+Streaming response (SSE)
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+When the ``stream`` parameter is set to ``true``, the server uses `Server-Sent
+Events
+<https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events>`__
+(SSE) to stream responses. Here is a detailed breakdown of the typical
+sequence and structure of events in a streaming response:
+
+* **HTTP Status**: 200 OK
+* **Content-Type**: text/event-stream
+* **Cache-Control**: no-cache
+
+The stream consists of a sequence of five events, each encapsulating part of
+the response in a structured format:
```

Comment:
I wonder if we should show the raw text of what the SSE looks like in real life rather than this slightly restructured text? Or do you think that linking to the MDN article is enough?

The things that are not terribly obvious here are:
1. What does an SSE event actually look like:

   ```
   event: message_start
   data: {"type":"message_start","message":{"id":"<message_id>","role":"assistant","model":"<model_name>"}}
   
   ```
2. You might receive the text one character at a time, or one whole event at a time, or several events at once, or the whole stream at once, so you'll need to do some work to buffer the messages and restructure them yourself. Maybe this part is a bit too much for these reference-level documentation since this really due to how chunked responses work over HTTP.

Reply:
I think it wouldn't be bad to show an example response. I'll add it.